Result/Result.swift

```diff
@@ -179,6 +179,15 @@ public func ?? <T, Error> (left: Result<T, Error>, @autoclosure right: () -> Res
 	return left.recoverWith(right())
 }
 
+// MARK: - Derive result from failable closure
+
+public func materialize<T>(f: () throws -> T) -> Result<T, ErrorType> {
```

Comment:
IMO, explicit specification is better than the alternative of having to later unwrap and cast the contained error. I also expect it will be common to do `return materalize(foo)` where the type info can be inferred from the function declaration where it's required.